### PR TITLE
[AutoTuner]Add memory model

### PR DESCRIPTION
**Pull request by Caozhou1995**
This PR adds memory model, which can be used to speed up pruning and filter out OOM memory strategies and strategies with low memory usage.

### CODE REVIEW

File: flagscale/auto_tuner/memory_model.py
Comment (aoyulong) [inline flagscale/auto_tuner/memory_model.py:1]: Please try to reuse this [impl](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/training/theoretical_memory_usage.py)

File: flagscale/auto_tuner/prune/memory.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            )
+            strategy["max_mem"] = "OOM"
+            strategy["performance"] = None
+            strategy["pruned"] = True
+            return True
+    return False
+
+
+def prune_by_memory_model_util(config, strategy, history=[]):
+    if "modeling_memory" in strategy:
```
Comment (aoyulong) [inline flagscale/auto_tuner/prune/memory.py]: It's better to rename "modeling_memory" to "memory_model" as the other places.

File: flagscale/auto_tuner/tuner.py [hunk 2]
```diff
@@ -78,16 +78,23 @@ def __init__(self, config: DictConfig):
...
+                    if "luvatar_BI" in os.environ["AIRS_ACCELERATOR_MODEL"]
+                    else int(os.environ["AIRS_ACCELERATOR_COUNT"])
                 )
                 # Set config
-                self.config.experiment.runner.nproc_per_node = int(
-                    os.environ["AIRS_ACCELERATOR_COUNT"]
+                self.config.experiment.runner.nproc_per_node = (
+                    int(os.environ["AIRS_ACCELERATOR_COUNT"]) * 2
+                    if "luvatar_BI" in os.environ["AIRS_ACCELERATOR_MODEL"]
+                    else int(os.environ["AIRS_ACCELERATOR_COUNT"])
```
Comment (aoyulong) [inline flagscale/auto_tuner/tuner.py]: Is there any way to move these platform related code into a standalone place? We may support differnt cloud platform.

File: flagscale/auto_tuner/memory_model.py
Comment (Caozhou1995) [reply flagscale/auto_tuner/memory_model.py:1]: The impl has been reused and the activation section has been refined.

File: flagscale/auto_tuner/prune/memory.py (hunk 1 above)
Comment (Caozhou1995) [reply flagscale/auto_tuner/prune/memory.py]: thx, done

File: flagscale/auto_tuner/tuner.py (hunk 2 above)
Comment (Caozhou1995) [reply flagscale/auto_tuner/tuner.py]: The platform code has been removed to platform.py and different platforms code will be in this file.

File: flagscale/train/theoretical_memory_usage.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
+"""
+Computes theoretical memory footprint for model training referring to megatron.
+Activation memory is optimized with adding block recompute formula.
```
Comment (aoyulong) [inline flagscale/train/theoretical_memory_usage.py:4]: Please add reference to megatron original impl

File: flagscale/auto_tuner/utils.py [hunk 4]
```diff
@@ -161,3 +167,114 @@ def compare_by_recompute(strategy1, strategy2):
         result = True
 
     return result
+
+
+def convert_config_to_megatron_args(config, strategy):
```
Comment (aoyulong) [inline flagscale/auto_tuner/utils.py:172]: Is there any simpler way to impl this?

**Review by aoyulong — APPROVED** [review]

**Review by aoyulong — APPROVED** [review]
LGTM

**Pull request merged**